File: ruby/boutique-inventory/boutique_inventory.rb
class BoutiqueInventory
  def initialize(items)
    @items = items
  end

  def item_names
    @items.map { |item| item[:name] }.sort
  end

  def cheap
    maximum_cost = 29.99
    @items.filter { |item| item[:price] <= maximum_cost }
  end

  def out_of_stock
    @items.filter { |item| item[:quantity_by_size] == {} }
  end

  def stock_for_item(name)
    @items.find { |item| item[:name] == name }[:quantity_by_size]
  end

  def total_stock
    @items.map { |item| item[:quantity_by_size].values.sum }.sum
  end

  private

  attr_reader :items
end

File: elixir/boutique-inventory/lib/boutique_inventory.ex
defmodule BoutiqueInventory do
  def sort_by_price(inventory) do
    Enum.sort_by(inventory, & &1.price)
  end

  def with_missing_price(inventory) do
    Enum.filter(inventory, &(&1.price == nil))
  end

  def update_names(inventory, old_word, new_word) do
    Enum.map(inventory, fn item ->
      %{item | name: String.replace(item.name, old_word, new_word)}
    end)
  end

  def increase_quantity(item, count) do
    increase_by_1 = fn {k, v} -> {k, v + count} end
    quantity_by_size = Map.new(item.quantity_by_size, increase_by_1)

    %{item | quantity_by_size: quantity_by_size}
  end

  def total_quantity(item) do
    Enum.reduce(item.quantity_by_size, 0, fn {_k, v}, acc -> acc + v end)
  end
end
